### PR TITLE
Pod Log Retrieval API & Validation Failure Log Preservation

### PR DESCRIPTION
**Pull request by paras-agarwal-meesho**
# 🔁 Pull Request Template – BharatMLStack

## Context:

> **Refer**: [Pod Log Retrieval API & Validation Failure Log Preservation](https://meesho.atlassian.net/wiki/spaces/EW/pages/5025923148/Pod+Log+Retrieval+API+Validation+Failure+Log+Preservation)

Model validation failures are currently difficult to debug because the Kubernetes pod logs of the test deployable are ephemeral and lost once the pod or deployable is recycled.

This solution introduces:
- A **standard Horizon API** to fetch Kubernetes pod logs for Argo CD applications.
- A **failure log preservation mechanism** that captures and uploads test deployable logs to GCS when validation fails.
- **Integration with TruffleBox** to surface preserved logs via GCS links in rejection reasons.

The goal is to eliminate dependency on direct Argo CD or cluster access and provide durable, auditable failure diagnostics.

## Describe your changes:
- API exposed: horizon/pkg/argocd/logs.go
- Tests for api: horizon/pkg/argocd/logs_test.go
- Validation Failure logs captured and uploaded to GCS: horizon/internal/predator/handler/predator_validation_logs.go
- Tests : horizon/internal/predator/handler/predator_validation_logs_test.go

## Testing:
- API testing was done
- Extensive test files for exhaustive testcases for all functions are created
- Flow is tested

## Checklist before requesting a review
- [x] I have reviewed my own changes?
- [x] Relevant or critical functionality is covered by tests?
- [ ] Monitoring needs have been evaluated?
- [x] Any necessary documentation updates have been considered?

## 📂 Modules Affected

- [x] `horizon` (Real-time systems / networking)
- [ ] `online-feature-store` (Feature serving infra)
- [ ] `trufflebox-ui` (Admin panel / UI)
- [ ] `infra` (Docker, CI/CD, GCP/AWS setup)
- [ ] `docs` (Documentation updates)
- [ ] Other: `___________`

---

## ✅ Type of Change

- [x] Feature addition
- [ ] Bug fix
- [ ] Infra / build system change
- [ ] Performance improvement
- [ ] Refactor
- [ ] Documentation
- [ ] Other: `___________`

---

## Summary by CodeRabbit

* **New Features**
  * New application-logs endpoint: fetch non-streaming logs with pod/container selection, time range, tailing, and filtering; maps backend errors to clear HTTP responses.
  * Validation workflow now captures failure logs, uploads them to cloud storage, and surfaces a log URI in validation results for easier troubleshooting.

* **Tests**
  * Extensive tests for log retrieval, NDJSON parsing, URL/query handling, validation log capture, time conversions, and numerous edge cases.

### CODE REVIEW

File: horizon/internal/infrastructure/controller/controller.go [hunk 1]
```diff
@@ -103,6 +123,96 @@ func (c *InfrastructureController) GetResourceDetail(ctx *gin.Context) {
...
+	if q.TailLines != "" {
+		var err error
+		tailLines, err = strconv.ParseInt(q.TailLines, 10, 64)
+		if err != nil || tailLines < 0 {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": gin.H{
+				"message": "tailLines must be a non-negative integer",
+			}})
+			return
+		}
+	}
```
[the remark this answers is not in the document]
Comment (paras-agarwal-meesho) [reply horizon/internal/infrastructure/controller/controller.go:187]: This is intended

File: horizon/internal/predator/handler/predator_validation_logs.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+// Prod: GcsConfigBucket
+// Non-prod: GcsModelBucket
+func (p *Predator) buildValidationLogsGCSPath(job *validationjob.Table) (bucket, objectPath string) {
+	if p.isNonProductionEnvironment() {
+		bucket = pred.GcsModelBucket
+		objectPath = path.Join(validationLogsDir, job.ServiceName, job.GroupID+".log")
+	} else {
+		bucket = pred.GcsConfigBucket
+		objectPath = path.Join(validationLogsDir, job.ServiceName, job.GroupID+".log")
+	}
```
[the remark this answers is not in the document]
Comment (paras-agarwal-meesho) [reply horizon/internal/predator/handler/predator_validation_logs.go:190]: Service Name does not contain "/" or ".."

File: horizon/internal/predator/handler/predator_validation_logs.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+	gcsURL := fmt.Sprintf("https://storage.cloud.google.com/%s/%s", bucket, objectPath)
+	log.Info().Str("gcsPath", gcsURL).Uint("jobID", job.ID).
+		Msg("Validation failure logs uploaded to GCS")
+	return gcsURL
+}
+
+// findDegradedPods queries the ArgoCD resource tree and returns only pods
+// whose health status is "Degraded".
+func (p *Predator) findDegradedPods(serviceName string) []infrastructurehandler.Node {
```
Comment (shubhamk-meesho) [inline horizon/internal/predator/handler/predator_validation_logs.go:89]: The pods which go into degraded state might be restarting repeatedly. Will the logs be available for older container then?

Comment (paras-agarwal-meesho) [reply horizon/internal/predator/handler/predator_validation_logs.go:89]: Yes, the Previous param in the api requests fetches the logs from the previous instance of the pod, this has been tested by using ArgoCD api.

**Comment by paras-agarwal-meesho** [comment]
Shifted to internal
